doc/holochain_101/src/json_string.md

```diff
@@ -0,0 +1,888 @@
+
+This is exactly like using `Into<String>` but for JSON data. An even looser
+option is to only require `TryInto<JsonString>` but this makes little or no
+difference to us in practise.
+
+An example of this is the `store_as_json` used to pass native Rust typed data
+across the WASM boundary. This is used internally by the `define_zome!` macro
+for all zome funtions:
+
+```rust
+pub fn store_as_json<J: TryInto<JsonString>>(
+    stack: &mut SinglePageStack,
+    jsonable: J,
+) -> Result<SinglePageAllocation, RibosomeErrorCode> {
+    let j: JsonString = jsonable
+        .try_into()
+        .map_err(|_| RibosomeErrorCode::ArgumentDeserializationFailed)?;
+
+    let json_bytes = j.into_bytes();
+    let json_bytes_len = json_bytes.len() as u32;
+    if json_bytes_len > U16_MAX {
+        return Err(RibosomeErrorCode::OutOfMemory);
+    }
+    write_in_wasm_memory(stack, &json_bytes, json_bytes_len as u16)
+}
+```
+
+The relevant `into()` or `try_into()` method is called _internally_ by the
+function accepting `Into<JsonString>`, meaning the caller needs to know almost
+nothing about _how_ the serialization is done. Additionally, the caller _could_
+do its own custom serialization, passing a `String` through, which would be
+wrapped as-is into a `JsonString`.
+
+Unfortunately this doesn't work as well for structs because of the way trait
+bounds work (or don't work) without complex boxing etc. See above for simple
+strategies to cope with nested/wrapped serialization in nested native data
+structures.
+
+This approach can be combined with the "quick and dirty" `Foo` with private
+`String` internals to create a `Foo` that can store _anything_ that round trips
+through `JsonString`:
+
+```rust
+struct Foo {
+  bar: String,
+}
+
+impl Foo {
+  fn new<J: Into<JsonString>> (bar: J) -> Foo {
+    Foo{ bar: String::from(JsonString::from(bar)) }
+  }
+
+  fn bar<T: TryFrom<JsonString>>(&self) -> Result<T, HolochainError> {
+    Ok(JsonString::from(self.bar.clone()).try_into()?)
+  }
+}
+
+// somewhere later..
+// we can build MyBar ad-hoc to send to Foo as long as it implements JsonString
+// we could create MyOtherBar in the same way and send to Foo in the same way
+#[derive(Serialize, Deserialize, Debug, DefaultJson)]
+struct MyBar { .. }
+
+let my_bar = MyBar::new(..);
+// auto stores as String via. JsonString internally
+let foo = Foo::new(my_bar);
+// note we must provide the MyBar type at restore time because we destroyed
+// that type info during the serialization process
+let restored_bar: MyBar = foo.bar()?;
+```
+
+This is how the `ContentAddressableStorage` trait used to work. It would
+"magically" restore the correct `Content` from storage based on an `Address`
+and type alone, provided the compiler had the type info available at compile
+time.
+
+We had to sacrifice this neat trick due to incompatible constraints put on the
+CAS by the reducer store etc. but it should work well in other scenarios :)
```

Comment:
„Reducer store?“
I think we had to choose to not use rust‘s type system for this so we can have dynamically linked CAS implementations.

Reply:
@lucksus yeah, how would you say that?